client/internal/retry/do.go

```diff
@@ -142,3 +144,84 @@ func Do(retryFunc Func, opt Options) error {
 		errors.Join(accumulatedErrs...),
 	)
 }
+
+func Do2(ctx context.Context, retryFunc Func, opt Options) error {
+	// set up context
+	ctxToUse := ctx
+	if opt.Timeout > 0 {
+		var cancel context.CancelFunc
+		ctxToUse, cancel = context.WithTimeout(ctx, opt.Timeout)
+		defer cancel()
+	}
+
+	// set up channels
+	retryChan := make(chan retryInfo)
+	go retry(ctxToUse, retryChan, opt.Delay, opt.Retries)
+	return doInternal(ctxToUse, retryFunc, retryChan, opt)
+}
+
+type retryInfo struct {
+	attempt int
+	timeout bool
+}
+
+// retry sends a retryInfo to the given channel according to given delay and retries,
+func retry(ctx context.Context, c chan<- retryInfo, delay time.Duration, retries int) {
+	c <- retryInfo{attempt: 0, timeout: false}
+	for attempt := 1; attempt <= retries || retries < 0; attempt++ {
+		if willTimeoutAfterDelay(ctx, delay) {
+			c <- retryInfo{attempt: attempt, timeout: true}
+		}
+		time.Sleep(delay)
+		c <- retryInfo{attempt: attempt, timeout: false}
```

Comment:
This can lead to a resource leak.

In the case that the function candidate for retrying is successful before `maxRetries - 2` this function will continue to execute until the for loop has completed. As the channel is not buffered - once we put another message the channel the goroutine will block as the channel is unable to receive any more messages. This will prevent the goroutine exiting and the channel closing.

Reply:
Discussed offline, thanks Chris 👍 Fixed